---
title: "Myanswers"
author: "Michael Blum"
date: "21 novembre 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", dev = "svg", fig.width = 5,
                      out.width = "70%", fig.asp = 0.7)
options(width = 110)
```

##Partie 1

###Question 1
The first column contains the ID of each individual, the second one contains the individual origin and the following ones contain the  allele counts.
```{r}
# Read the data
load(url("http://membres-timc.imag.fr/Michael.Blum/geno_pheno_BEE3.Rdata"))
# Print the genotype for the first 10 markers and the first 300 indiv. in the data matrix
genotype[c(1, 300), 1:8]
```


###Question 2
```{r}
# Use bigstatsr
library(bigstatsr)
(geno <- big_copy(as.matrix(genotype[, -(1:2)])))
# Perform PCA
obj.svd <- big_SVD(geno, big_scale())
PCs <- obj.svd$u[, 1:3]
# Plot the PC scores of the 1st PC
library(ggplot2)
origin <- as.character(genotype[, 2])
qplot(rows_along(geno), PCs[, 1], color = origin) +
  labs(x = "Index of individual", y = "PC1", color = "Origin")
```

### Simulate trait1 and trait2
```{r}
n <- nrow(geno)
p <- ncol(geno)
lpval.thr <- log10(0.05) 
nsimu <- 100
M <- 10
h2vec <- seq(0.01, 0.99, length = nsimu)
pval_5 <- rep(NA_real_, nsimu)

for (i in seq_along(h2vec)) {
  
  cat("Heritability :", h2 <- h2vec[i], "\n")
  
  env <- rnorm(n, sd = sqrt((1 - h2) / 2))
  causal <- sample(p, size = 2 * M, replace = FALSE)
  betas1 <- rnorm(M, sd = sqrt(h2 / M))
  betas2 <- rnorm(M, sd = sqrt(h2 / M))
  trait1 <- scale(geno[, causal[1:M]]) %*% betas1 + 
    env + rnorm(n, sd = sqrt((1 - h2) / 2))
  trait2 <- scale(geno[, causal[M + 1:M]]) %*% betas2 +
    env + rnorm(n, sd = sqrt((1 - h2) / 2))
  
  # summary(lm(trait1 ~ trait2))
  
  gwas <- big_univLinReg(X = geno, y.train = trait1, covar.train = PCs)
  lpval <- predict(gwas)
  
  SNPs_5 <- (lpval < lpval.thr)
  trait1g <- geno[, SNPs_5, drop = FALSE] %*% gwas$estim[SNPs_5]
  
  pval_5[i] <- summary(lm(trait2 ~ trait1g + PCs))$coefficient[2, 4]
}
```


```{r}
qplot(h2vec, log10(pval_5)) +
  geom_hline(yintercept = lpval.thr, linetype = 2, color = "red") +
  geom_smooth(method = "loess") +
  labs(x = "Heritability", y = "log10(p-value)")
```

